WordPress/wordpress-develop
# Allow patterns files to be organized in sub-folders.

Trac ticket: https://core.trac.wordpress.org/ticket/62378
Gutenberg issue: https://github.com/WordPress/gutenberg/issues/50144

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

**Description**

I'm changing the initial `$dirpath` to avoid having the trailing slash, as the file exist can behave differently in different file systems.

Then, I'm making use of the static method `scandir` to check for php files in the directory recursively. Finally, adding the trailing slash so that this [str_replace](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-theme.php#L1961) leaves only the required slashes in the `$pattern_data` array.

**Testing Instructions**

1. Use a block theme.
2. Add a pattern in your theme, for example, `/patterns/subfolder/my-pattern.php`
3. Open a post or page.
4. Confirm that 'my-pattern' shows up
